toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -411,6 +444,96 @@ func WaitForLoopbackToDetach(devicePath string, diskPath string) error {
 }
 
+func WaitForDiskDevice(diskDevPath string) error {
+	err := WaitForDevicesToSettle()
+	if err != nil {
+		return err
+	}
+
+	// 'udevadm settle' is sometimes not enough.
+	// So, double check that the partitions have been populated.
+	// Ideally, we would use 'udevadm wait' instead of 'udevadm settle'. But it is too new and so isn't universally
+	// available yet.
+	err = waitForDiskToPopulate(diskDevPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func waitForDiskToPopulate(diskDevPath string) error {
+	partitionTable, err := ReadDiskPartitionTable(diskDevPath)
+	if err != nil {
+		return err
+	}
+
+	if partitionTable == nil {
+		// Disk is empty.
+		return nil
+	}
+
+	_, err = retry.RunWithExpBackoff(context.Background(), func() error {
+		kernelPartitions, err := GetDiskPartitions(diskDevPath)
+		if err != nil {
+			return err
+		}
+
+		errs := []error(nil)
+		for _, partition := range partitionTable.Partitions {
+			info, found := sliceutils.FindValueFunc(kernelPartitions, func(info PartitionInfo) bool {
+				return info.Path == partition.Path
+			})
+			if !found {
+				err := fmt.Errorf("failed to find partition device node (%s)", partition.Path)
+				errs = append(errs, err)
+				continue
+			}
+
+			if !strings.EqualFold(partition.PartTypeUuid, info.PartitionTypeUuid) {
+				err := fmt.Errorf("partition's (%s) type UUID is wrong: expected (%s), actual (%s)",
+					partition.Path, partition.PartTypeUuid, info.PartitionTypeUuid)
+				errs = append(errs, err)
+			}
+
+			if !strings.EqualFold(partition.PartUuid, info.PartUuid) {
+				err := fmt.Errorf("partition's (%s) UUID is wrong: expected (%s), actual (%s)",
+					partition.Path, partition.PartUuid, info.PartUuid)
+				errs = append(errs, err)
+			}
+
+			if partition.PartLabel != info.PartLabel {
+				err := fmt.Errorf("partition's (%s) label is wrong: expected (%s), actual (%s)",
+					partition.Path, partition.PartLabel, info.PartLabel)
+				errs = append(errs, err)
+			}
+
+			if partition.FileSystemType != info.FileSystemType {
+				err := fmt.Errorf("partition's (%s) filesystem type is wrong: expected (%s), actual (%s)",
+					partition.Path, partition.FileSystemType, info.FileSystemType)
+				errs = append(errs, err)
+			}
+
+			if !strings.EqualFold(partition.FileSystemUuid, info.Uuid) {
+				err := fmt.Errorf("partition's (%s) filesystem UUID is wrong: expected (%s), actual (%s)",
+					partition.Path, partition.FileSystemUuid, info.Uuid)
+				errs = append(errs, err)
+			}
+		}
+
+		if len(errs) > 0 {
+			return errors.Join(errs...)
+		}
+
+		return nil
+	}, 10, 120*time.Millisecond, 2.0)
+	if err != nil {
+		return fmt.Errorf("timed out waiting for disk (%s) info to be populated:\n%w", diskDevPath, err)
+	}
+
+	return nil
+}
+
 // WaitForDevicesToSettle waits for all udev events to be processed on the system.
 // This can be used to wait for partitions to be discovered after mounting a disk.
 func WaitForDevicesToSettle() error {
```

Comment:
Do we see other usages for `WaitForDevicesToSettle()` or should we remove it to prevent accidental usage in scenarios where it is not behaving accurately?
I still see a reference in https://github.com/microsoft/azure-linux-image-tools/blob/9d4d908f0cf2a969f1ae2d3857d1523714e06115/toolkit/tools/pkg/imagecustomizerlib/customizepartitionsuuids.go#L68

Reply:
Good call.